Add String format: byte pattern support

aeoncl opened the pull request:
**What**: Added support for OpenAPI string byte pattern

**Why**: Sometimes in life, you encounter an openAPI spec with a string byte property:

```yaml
myString:
  type: string
  format: byte
```

**How**:

We parsed the OpenAPI ByteArrayPattern to Specmatic's BinaryPattern

**Checklist**:

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation N/A
- [X] Tests
- [ ] Sonar Quality Gate N/A

**Issue ID**:
Closes: /

Snorkell committed:
Handle byte array string as binarySchema

Snorkell committed:
bad test to be refactor

aeoncl committed:
Fix string byte part test response body

aeoncl committed:
Revert "Fix string byte part test response body"

This reverts commit eef9b394707392e79f947396c8b755274d21dc63.

aeoncl committed:
Revert "bad test to be refactor"

This reverts commit 3ee3199ee6cab44b8deb7c9fb5dac6c29650b171.

aeoncl committed:
Fix string byte part test response body

harikrishnan83 commented on core/src/test/kotlin/in/specmatic/conversions/OpenApiSpecificationTest.kt:
```diff
@@ -6467,6 +6468,63 @@ paths:
+                            ), HttpResponse.OK("{\"filename\": \"ThIsi5ByT3sD4tA\"}")
+                    )
+
+            println(result.reportString())
```
Can we please remove the print statement?

aeoncl committed:
Removed unused print statement

joelrosario commented on core/src/main/kotlin/in/specmatic/conversions/OpenApiSpecification.kt:
```diff
@@ -627,6 +627,7 @@ class OpenApiSpecification(private val openApiFile: String, val openApi: OpenAPI
                     throw UnsupportedOperationException("Unsupported composed schema: $schema")
                 }
             }
+            is ByteArraySchema -> BinaryPattern()
```
We'd need to implement a new Pattern class for ByteArraySchema here. See my comment at https://github.com/znsio/specmatic/pull/716#issuecomment-1722482596.

aeoncl replied:
Yes, i came to the same conclusion while working with this quick fix: it needs a new Pattern with it's own matching logic & generation.

It would be cool to be able to use externalValue for stubbing with this Pattern Class (taking the file bytes and encoding them to base64)
Currently if i understand correctly: 

When dealing with externalValue , the parser adds _filename to the key that corresponds to the file path, then somewhere in the response generation, the file content is added to the response without looking at the pattern type.

Maybe we'll need to adapt this behaviour to support different externalValue treatment based on the PatternType?

aeoncl replied:
After looking further, i realized that externalValue is not supported for the responses yet. i did not find anything that loads file content in the response. This could be useful if we want to send a string of bytes or just return an application octet stream, i'm gonna create an issue for this

joelrosario replied:
@aeoncl that's right, Specmatic does not yet support for `externalValue`.

We have been using examples in the specification file, as we find that keeping the example inline tends to make it easier to understand the request format.

However I also see what you mean and I can see that `externalValue` ([documented here](https://swagger.io/docs/specification/adding-examples/)) could be useful if you need to use a base64 encoded file as an example.

aeoncl committed:
Add Base64StringPattern

joelrosario committed:
Merge branch 'main' into add-string-byte-pattern-2

joelrosario committed:
Added StringPattern as negative type

For a string that has to be base64-encoded, a non-base64 encoded string
is a negative example

joelrosario committed:
Base64StringPattern.parse should not encode a non-base64 value

The intent of this method is only to convert a string to the right
data-type. This usually involves an implicit format check. For example,
a value like "pqr" cannot be parsed to a decimal number, and an
exception would result.

When the target type is a base64 string (which is actually a format), do
the necessary validation without changing the value being parsed.

joelrosario reviewed (changes requested):
Hi @aeoncl, thanks, looks good!

I've added a test and slightly modified a line of code. I've given reasons for this in the respective commit messages. Hope this is fine.

Could you add 2 integration tests:
1. a test with a specification containing a byte string with an base64 example running as test (see tests in the `AuthorizationCodeFlowTest` class for ideas on how to do this)
2. a test with a specification containing a byte string with an expectation running as stub (see the test named `should create stub from gherkin that includes OpenAPI spec` for an example of how you could do this)

joelrosario committed:
Added tests and updated Base64StringPattern.negativeBasedOn

joelrosario committed:
Merge branch 'main' into add-string-byte-pattern-2

joelrosario reviewed (approved):
Thanks for the PR @aeoncl and for your patience. I've added some tests, and made a few minor changes to Base64StringPattern.

Merging this PR.
